### PR TITLE
FIX: Do not allow , or . in site setting integer input

### PR DESCRIPTION
**Pull request by martin-brennan**
Followup to e113eff6634c9b398b0541ea6a0d1626f6732dcc

We previously sanitized input for integer site settings
on the server side only, which was a bit confusing when
users would enter e.g. 100.5 and end up with 1005, and
not see this reflected in the UI.

Now that we are using native number inputs for these settings,
we can improve the experience a bit by not allowing `.` or `,`
in the input, because it should be whole numbers only, and
add a step size of 1. All other characters are already prevented
in this native number input.

### CODE REVIEW

**Review by Drenmi — APPROVED** [review]
💯.0

File: app/assets/javascripts/discourse/tests/integration/components/site-setting-test.js [hunk 1]
```diff
@@ -88,4 +88,16 @@ module("Integration | Component | site-setting", function (hooks) {
...
+
+  test("prevents decimal in integer setting input", async function (assert) {
+    this.set("setting", {
+      setting: "suggested_topics_unread_max_days_old",
+      value: "90",
+      type: "integer",
+    });
+
+    await render(hbs`<SiteSetting @setting={{this.setting}} />`);
+    await fillIn(query(".input-setting-integer"), "90.5");
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/tests/integration/components/site-setting-test.js:100]: Is `fillIn` "typing" the characters one by one or setting the `value` of the input? 🤔

If it's the former, then I would expect the dot to be "swallowed" and the result to be 905.

If it's the latter, are we really testing your change or just the behaviour provided by html when using "step=1"?

Comment (CvX) [reply app/assets/javascripts/discourse/tests/integration/components/site-setting-test.js:100]: [`fillIn`](https://github.com/emberjs/ember-test-helpers/blob/master/API.md#fillin) sets the value all at once.
[`typeIn`](https://github.com/emberjs/ember-test-helpers/blob/master/API.md#typein) is emulating all the steps.
I believe we should use `typeIn` here.

Comment (martin-brennan) [reply app/assets/javascripts/discourse/tests/integration/components/site-setting-test.js:100]: Oh yes you are both right, I will use typeIn, I didn't realise this existed and have previously been messing around with our triggerKeyPress code

**Review by CvX — COMMENTED** [review]
(oops, had two comments in a pending state…)

File: app/assets/javascripts/admin/addon/components/site-settings/integer.gjs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+      event.preventDefault();
+      event.stopPropagation();
+      return;
+    }
+  }
+
+  <template>
+    <input
+      {{on "keydown" this.preventDecimal}}
+      {{on "input" (withEventValue this.updateValue)}}
```
[the remark this answers is not in the document]
Comment (martin-brennan) [reply app/assets/javascripts/admin/addon/components/site-settings/integer.gjs:32]: Ah okay cool, will do that instead :+1:

**Comment by martin-brennan** [comment]
@CvX I changed to `typeIn` with https://github.com/discourse/discourse/pull/27618/commits/d378598d0a4c08ffd032aadfcc475aef49bfdc9e but still getting weird issues in the test. I set a `delay: 1000` for `typeIn` to see what is happening easier, and though the `.` is prevented correctly, the value of the input seems to be reset to `''` before `5` is typed, which results in the final value of the input being `5`. See the sequence of events below:

image

This does not happen when I type manually in the UI, so I feel like maybe it's something to do with how `typeIn` fires the events? From the docs:

> The major difference between typeIn and fillIn is that typeIn triggers keyboard events as well as input and change. Typically this looks like focus -> focusin -> keydown -> keypress -> keyup -> input -> change per character of the passed text (this may vary on some browsers).

Could you please have a look at this and see if there is some other way this can be tested, or something I have gotten wrong?

**Comment by martin-brennan** [comment]
@CvX any luck with the test here? If not I think I should just get rid of the test and merge it, the behaviour works fine outside the test environment

**Comment by CvX** [comment]
Yeah, ember-test-helpers doesn't check for `defaultPrevented` when firing that event chain. I'd like to fix that but in the meantime go ahead and mark this test as `skipped` 👍

**Pull request merged**